### PR TITLE
change latest version ID on crate to match CrateDetails.latest_version

### PR DESCRIPTION
**Pull request by syphar**
in all release-lists the one version we show should be the latest version (non-yanked and non-prerelease) to match the logic in the rustdoc page header. Otherwise we are showing releases in the lists that directly lead to a page with a _go to latest version_ link. 
Now `crates.latest_version_id` is calculated with the same logic and updated after each build. 

This is a preparation for #708, which I would do when #1521 is merged. Then it's just switching the calculated subquery to just using `crates.latest_version_id`. 

The data migration is a very simple impl, and takes 15 minutes on my machine (with macos virtualized docker). IMHO it's fine running that long on prod, of course I can improve speed when needed.

### CODE REVIEW

**Comment by jyn514** [comment]
@Nemo157 would you mind helping review this?

File: src/web/crate_details.rs [hunk 1]
```diff
@@ -233,10 +232,14 @@ impl CrateDetails {
     }
 }
 
-fn releases_for_crate(conn: &mut Client, crate_id: i32) -> Vec<Release> {
+fn releases_for_crate(
+    conn: &mut impl GenericClient,
+    crate_id: i32,
+) -> Result<Vec<Release>, postgres::Error> {
```
Comment (Nemo157) [inline src/web/crate_details.rs:238]: These functions should all use `anyhow::Error` to get a backtrace attached as soon as possible.

File: src/web/crate_details.rs [hunk 2]
```diff
@@ -76,11 +77,11 @@ pub struct Release {
 
 impl CrateDetails {
     pub fn new(
-        conn: &mut Client,
+        conn: &mut impl GenericClient,
```
Comment (Nemo157) [inline src/web/crate_details.rs:80]: Was gonna say this should probably be `conn: impl GenericClient`, but `postgres` is missing the `impl GenericClient for &mut impl GenericClient` forwarding impl 😭

File: src/db/add_package.rs [hunk 3]
```diff
@@ -127,12 +128,15 @@ pub(crate) fn add_package_into_database(
     add_keywords_into_database(conn, metadata_pkg, release_id)?;
     add_compression_into_database(conn, compression_algorithms.into_iter(), release_id)?;
 
-    // Update the crates table with the new release
+    let crate_details = CrateDetails::new(conn, &metadata_pkg.name, &metadata_pkg.version, None)
+        .with_context(|| anyhow!("error when fetching crate-details"))?
```
Comment (Nemo157) [inline src/db/add_package.rs:132]: This can be just `.context("error when fetching crate-details")`

File: src/web/crate_details.rs (hunk 1 above)
Comment (syphar) [reply src/web/crate_details.rs]: I didn't know about losing the backtrace here. 

I changed the return types, even though by doing that I had to switch to panicking in the data migration in `db::migrate`.

File: src/web/crate_details.rs [hunk 4]
```diff
@@ -470,48 +470,56 @@ mod tests {
                         build_status: true,
                         yanked: false,
                         is_library: true,
+                        id: details.releases[0].id,
```
[the remark this answers is not in the document]
Comment (syphar) [reply src/web/crate_details.rs:473]: both `assert_matches!` and `const {}` look awesome :) 

Question is if we want to give up stable support for this. 

( Actually I didn't check what runs in prod? )

**Review by Nemo157 — APPROVED** [review]
LGTM, tested locally and appears to behave correctly.

I assume for the migration we just need to stop building, {update server, run migration} (in parallel/either order), then start building again?

**Comment by syphar** [comment]
> LGTM, tested locally and appears to behave correctly.
awesome, thanks! 

> I assume for the migration we just need to stop building, {update server, run migration} (in parallel/either order), then start building again?

yep, that would be one way. 

Another way would be so move the data migration after the deploy/restart. Since we're only updating data and no schema, this would be fine. 

The migration is not really optimized, so it will take some time (locally it were 10 minutes) 


I'll fix the conflicts with master, then we can merge.

File: src/db/add_package.rs [hunk 5]
```diff
@@ -127,12 +128,21 @@ pub(crate) fn add_package_into_database(
     add_keywords_into_database(conn, metadata_pkg, release_id)?;
     add_compression_into_database(conn, compression_algorithms.into_iter(), release_id)?;
 
-    // Update the crates table with the new release
+    let crate_details = CrateDetails::new(
+        conn,
+        &metadata_pkg.name,
+        &metadata_pkg.version,
+        &metadata_pkg.version,
```
Comment (syphar) [inline src/db/add_package.rs:135]: @jyn514 @jsha without me digging into #1527, 
is it correct just to use the version twice?

Comment (jyn514) [reply src/db/add_package.rs:135]: Yeah that's fine, it only matters for the frontend templates.

File: src/db/migrate.rs [hunk 6]
```diff
@@ -5,19 +5,8 @@ use postgres::{Client, Error as PostgresError, Transaction};
...
-/// Downgrade query should return database to previous state.
-///
-/// Example:
-///
-/// ```
-/// let my_migration = migration!(100,
-///                               "Create test table",
-///                               "CREATE TABLE test ( id SERIAL);",
-///                               "DROP TABLE test;");
-/// ```
```
[the remark this answers is not in the document]
Comment (syphar) [reply src/db/migrate.rs:18]: I'm not sure any more, I'll add it back

**Comment by syphar** [comment]
This seems to be tested and approved now, 

I'll merge it when I have time to also deploy it, so this migration is not blocking anyone.

**Pull request merged**